File: 198_House_Robber/houseRobber.cpp
class Solution
{
public:
    int rob(vector<int> &nums)
    {

        int len = nums.size();

        // Padding
        vector<int> re(len + 2, 0);

        // 自底向上
        for (int i = len - 1; i >= 0; i--)
        {
            re[i] = std::max(re[i + 1], nums[i] + re[i + 2]);
        }

        return re[0];
    }
};

File: 198_House_Robber/houseRobber.py
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return 0
        
        take, not_take = nums[0], 0

        for i in range(1, len(nums)):
             take, not_take =  not_take + nums[i], max(take, not_take)

        return max(take, not_take)
